### ISSUE TITLE
Multipart payload image broken when uploading through hapi

### ISSUE THREAD

**Issue by AndreiLucaci** [hapijs/hapi#4153]
#### Support plan

* *is this issue currently blocking your project?* (yes/no): yes: we cannot process input image files
* *is this issue affecting a production system?* (yes/no): yes: we cannot deploy a stable version in production without this

#### Context

* *node version*: v12.18.3
* *module version with issue*: 19.1.1
* *last module version without issue*: ?
* *environment* (e.g. node, browser, native): node / native
* *used with* (e.g. hapi application, another framework, standalone, ...): standalone
* *any other relevant information*:

#### What are you trying to achieve or the steps to reproduce?

I have set up a `POST` route to receive a payload of type:

```js
 payload: {
        parse: true,
        multipart: {
          output: 'file',
        },
        output: 'file',
        allow: ['multipart/form-data', 'image/jpeg', 'image/png'],
        uploads: 'uploads',
  } as RouteOptionsPayload,
```

and a handler as such:
```js
testUpload: async (req, h) => {
    const { payload } = req;
    // the payload.file.path on disk is corrupt

    const binary = fs.readFileSync(payload.file.path, 'binary'); // this payload.file.path on disk is already corrupt
    const base64 = new Buffer(binary, 'binary').toString('base64');

    fs.writeFileSync('out.png', base64, 'base64'); // corrupt image

    return h.file('out.png').header('content-type', payload.file.headers['content-type']); // corrupt response
},
```

and here's the CURL:
```
curl --location --request POST 'http://localhost:4017/v1/test-upload' \
--form 'file=@/Users/andreilucaci/Downloads/dbl8xu5-49b389f4-8814-49ff-adaf-06c70d60bdfd.jpg_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1cm46YXBwOiIsImlzcyI6InVybjphcHA6Iiwib2JqIjpbW3sicGF0aCI6IlwvZlwvYmI1OGU0MDctZTNiYi00YzQwLTkzNjYtOTQ2ZGFmOGI3N2ZhXC9kYmw4eHU1LTQ5YjM4OWY0.png'
```

The issue is that the file that is stored (temporarily) on the disk is already corrupted. I don't know exactly why this happens.

The issue is larger, since this is just a demo, cut out from an S3 photo integration service, which is broken, since every photo that is uploaded is broken!

#### What was the result you got?
Corrupt image / data

#### What result did you expect?
Valid image to be stored!
Valid image to be returned!

**Comment by kanongil** [comment]
This is not a hapi issue. You are doing a lossy 'binary' encoding of your file read, without any way to recover.

**Comment by AndreiLucaci** [comment]
> This is not a hapi issue. You are doing a lossy 'binary' encoding of your file read, without any way to recover.

You missed the point. The `binary` part was just an example. The file that hapi stores as a temporary file, in the uploads folder, is already broken. Something that hapi does with the request and with the file makes it impossible to work with file upload.

If you have read the original issue, you would have seen that the issue is the corrupt binary stream that hapi provides, be is as a `file`, `data` or `stream`.

This is not fixed, and this is an issue with how hapi handles the request!

**Comment by kanongil** [comment]
Sorry, you have provided an unreproducible case, where your code corrupts any file it reads, so it seems pretty open and shut.

Hapi uploads normally works, so you will need to provide more details for us to reproduce the issue, and investigate further.

**Comment by devinivy** [comment]
It may not be obvious, but the "binary" encoding in your example @AndreiLucaci just means the latin1 character set, which is very limited and therefore you lose information while you're reading the file.  I suggest reading the file directly into a buffer by omitting an encoding in your call to `readFileSync()`.

**Comment by AndreiLucaci** [comment]
Hello

Ok, just ignore the `'binary'` part and let's do this again.

And please, excusing my ignorance, help me understand something:
I have this route configuration:
```ts
payload: {
        parse: true,
        multipart: {
          output: 'file',
        },
        output: 'file',
        allow: ['multipart/form-data', 'image/jpeg', 'image/png'],
        uploads: 'uploads',
  } as RouteOptionsPayload,
```

and this request:
```
curl --location --request POST 'http://localhost:4017/v1/test-upload' \
--form 'file=@/Users/andreilucaci/Downloads/dbl8xu5-49b389f4-8814-49ff-adaf-06c70d60bdfd.jpg_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1cm46YXBwOiIsImlzcyI6InVybjphcHA6Iiwib2JqIjpbW3sicGF0aCI6IlwvZlwvYmI1OGU0MDctZTNiYi00YzQwLTkzNjYtOTQ2ZGFmOGI3N2ZhXC9kYmw4eHU1LTQ5YjM4OWY0.png'
```

then the temporary file will be stored in the `'uploads'` folder.
Is the temporary file supposed to be altered (e.x. encoded) from the original file that was inputed?
e.g.
I'm uploading `dbl8xu5-49b389f4-8814-49ff-adaf-06c70d60bdfd.jpg`, and I have this:

| filename | size (in B) | description| location  |
|---------|---------|-------|----------|
|dbl8xu5-49b389f4-8814-49ff-adaf-06c70d60bdfd.jpg|159498|original|/.../Downloads/dbl8xu5-49b389f4-8814-49ff-adaf-06c70d60bdfd.jpg|
|1598863227424-1234-74e2efaa2e9be9e5|238398| temp file|/.../uploads/1598863227424-1234-74e2efaa2e9be9e5|

So, the temp file is larger in B from the original file.  Why is that? Does hapi do something with the file? Is my route configuration for the payload incorrect? 

Andrei

**Comment by AndreiLucaci** [comment]
Hi all,

So, I figured out the issue. Hapi is `UTF8` encoding the files, which results in corrupted data streams if not interpret carefully, and we know how JS handles I/O stream as strings, not being binary safe...

TL;DR
So the solution to my problem is:

```ts
    payload: {
        multipart: {
          output: 'file',
        },
        parse: true,
        output: 'file',
        allow: ['multipart/form-data', 'image/jpeg', 'image/png'],
        uploads: 'uploads',
      } as RouteOptionsPayload,
```

and in the handler:
```js
// ...
fs.createReadStream(payload.file.path, { encoding: 'utf8' }).pipe(fs.createWriteStream('out.jpeg', { encoding: 'binary' }));
```
